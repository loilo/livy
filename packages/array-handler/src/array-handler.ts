import type {
  LogRecord,
  ResettableInterface,
  SyncHandlerInterface,
} from '@livy/contracts'
import { AbstractLevelBubbleHandler } from '@livy/util/handlers/abstract-level-bubble-handler'
import { MirrorSyncHandlerMixin } from '@livy/util/handlers/mirror-sync-handler-mixin'
import { ProcessableHandlerMixin } from '@livy/util/handlers/processable-handler-mixin'

/**
 * Stores log records in an array; great for debugging
 */
export class ArrayHandler
  extends MirrorSyncHandlerMixin(
    ProcessableHandlerMixin(AbstractLevelBubbleHandler),
  )
  implements SyncHandlerInterface, ResettableInterface
{
  private _records: LogRecord[] = []

  /**
   * @inheritdoc
   */
  public handleSync(record: LogRecord) {
    this._records.push(record)
    return !this.bubble
  }

  /**
   * @inheritdoc
   */
  public handleBatchSync(records: LogRecord[]) {
    this._records.push(...records)
  }

  /**
   * Get the stored records
   */
  public get records() {
    return this._records
  }

  /**
   * @inheritdoc
   */
  public reset() {
    this._records.splice(0)
  }
}
