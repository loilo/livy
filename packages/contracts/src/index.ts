export * from './closable-handler-interface.js'
export * from './formattable-handler-interface.js'
export * from './formatter-interface.js'
export * from './handler-interface.js'
export * from './log-level.js'
export * from './log-record.js'
export * from './logger-interface.js'
export * from './processable-handler-interface.js'
export * from './processor-interface.js'
export * from './resettable-interface.js'
