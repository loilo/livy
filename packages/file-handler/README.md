# `@livy/file-handler`

This [Livy](../../README.md#readme) handler writes log records to a file.

---

[**Synchronous logger support:**](../../README.md#synchronous-and-asynchronous-logging) yes

**Runtime:** Node.js

---

## Basic Example

```js
import { FileHandler } from '@livy/file-handler'

const handler = new FileHandler('logs.txt')
```

## Installation

Install it via npm:

```bash
npm install @livy/file-handler
```

## Options

The first argument to this handler's constructor is a file path to write to.

An object of options can be passed to the constructor as the second argument.

The following options are available:

### `append`

**Type:** `boolean`

**Default:** `true`

**Description:** Whether DOM elements should be added at the beginning of the container instead of the end

### `bubble`

**Type:** `boolean`

**Default:** `true`

**Description:** Controls whether records handled by this handler should bubble up to other handlers.

> See also: [Bubbling](../../README.md#bubbling)

### `formatter`

**Type:** [`FormatterInterface`](../contracts/README.md#formatterinterface)

**Default:** [`new LineFormatter()`](../line-formatter/README.md#readme)

**Description:** The formatter to use.

### `level`

**Type:** [`LogLevel`](../contracts/README.md#loglevel)

**Default:** `'debug'`

**Description:** Controls which log records should be handled based on their [log level](../../README.md#log-levels).

### `prefix`

**Type:** `string | string[]`

**Default:** `[]`

**Description:** Prepend one or more lines to the file when it's first created. This can be useful to prepend header data, like in CSV.

## Public API

### `bubble`

Controls whether records handled by this handler should bubble up to other handlers. Initially set through the [`bubble`](#bubble) option.

> See also: [Bubbling](../../README.md#bubbling)

### `close()`

This handler implements the [`ClosableHandlerInterface`](../contracts/README.md#closablehandlerinterface). On cleanup, it closes the internal file handler.

> You usually don't want to call this method manually. It is done automatically when a Node.js process exits / a browser page is closed.

### `defaultFormatter` (read-only)

The formatter used by this handler if no [`formatter`](#formatter) option is set.

### `formatter`

This handler supports formatters by implementing the [`FormattableHandlerInterface`](../contracts/README.md#formattablehandlerinterface).

### `level`

The minimum [log level](../../README.md#log-levels) of a log record to be considered by this handler. Initially set through the [`level`](#level) option.

### `processors`

This handler supports [processors](../../README.md#processors) by implementing the [`ProcessableHandlerInterface`](../contracts/README.md#processablehandlerinterface).

### `reset()`

This handler implements the [`ResettableInterface`](../contracts/README.md#resettableinterface). Resetting it resets all attached processors.

> You usually don't want to call this method manually on an individual handler. Consider calling it [on the logger](../logger/README.md#reset) instead.
