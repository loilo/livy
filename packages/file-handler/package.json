{
  "name": "@livy/file-handler",
  "type": "module",
  "version": "1.0.4",
  "description": "Writes Livy log records to a file",
  "keywords": [
    "logging",
    "logger",
    "logs",
    "log",
    "loglevel",
    "tools",
    "monolog",
    "handler",
    "file",
    "logfile"
  ],
  "author": "loilo <florian.reuschel@gmx.de>",
  "homepage": "https://github.com/loilo/livy/tree/master/packages/file-handler#readme",
  "license": "MIT",
  "exports": {
    "types": "./lib/file-handler.d.ts",
    "default": "./lib/file-handler.js"
  },
  "sideEffects": false,
  "directories": {
    "lib": "lib",
    "test": "test"
  },
  "files": [
    "lib"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": "https://github.com/loilo/livy/tree/master/packages/file-handler",
  "scripts": {
    "build": "del lib && tsc --declaration",
    "lint:readme": "remark --quiet --no-stdout -u validate-links README.md",
    "lint:source": "eslint \"src/**/*.ts\"",
    "test": "vitest run --coverage"
  },
  "bugs": {
    "url": "https://github.com/loilo/livy/issues"
  },
  "dependencies": {
    "@livy/contracts": "^1.1.2",
    "@livy/util": "^1.0.6",
    "@types/node": "^18.11.18"
  },
  "devDependencies": {
    "typescript": "^5.1.0",
    "@livy/bin": "file:../../bin",
    "@livy/test-utils": "file:../../test-utils"
  },
  "gitHead": "bd96a3e4d1e5870873b0e0b9e538486aea6343a2"
}
