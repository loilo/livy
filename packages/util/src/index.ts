export type * from './formatters/abstract-batch-formatter.js'
export type * from './formatters/included-record-properties.js'
export type * from './formatters/line-formatter.js'
export type * from './handlers/abstract-batch-handler.js'
export type * from './handlers/abstract-formatting-processing-handler.js'
export type * from './handlers/abstract-level-bubble-handler.js'
export type * from './handlers/formattable-handler-mixin.js'
export type * from './handlers/is-closable-handler-interface.js'
export type * from './handlers/is-formattable-handler-interface.js'
export type * from './handlers/is-processable-handler-interface.js'
export type * from './handlers/is-sync-handler-interface.js'
export type * from './handlers/mirror-sync-handler-mixin.js'
export type * from './handlers/processable-handler-mixin.js'
export type * from './handlers/respect-level-mixin.js'
export type * from './console-adapter-interface.js'
export type * from './console-adapter.js'
export type * from './environment.js'
export type * from './gated-set.js'
export type * from './helpers.js'
export type * from './html.js'
export type * from './is-resettable-interface.js'
export type * from './mixin.js'
export type * from './timer.js'
export type * from './types.js'
export type * from './validatable-set.js'
