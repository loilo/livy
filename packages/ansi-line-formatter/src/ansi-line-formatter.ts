import type { LogLevel } from '@livy/contracts'
import { LineFormatter, LineFormatterOptions } from '@livy/line-formatter'
import { isEmpty } from '@livy/util/helpers'
import chalk, { supportsColor } from 'chalk'
import { DateTime } from 'luxon'

import { emphasize } from 'emphasize'
import chalkTemplate from 'chalk-template'

export interface AnsiLineFormatterOptions extends LineFormatterOptions {
  /**
   * Whether records should be highlighted with ANSI escape sequences
   * If not provided (or set to `undefined`), color support is going
   * to be automatically detected
   */
  decorated: boolean | undefined
}

/**
 * Formats log records as single lines with terminal highlighting
 */
export class AnsiLineFormatter extends LineFormatter {
  /**
   * Whether records should be highlighted with ANSI escape sequences
   */
  public decorated: boolean | undefined

  public constructor({
    decorated,
    ...options
  }: Partial<AnsiLineFormatterOptions> = {}) {
    super(options)
    this.decorated = decorated
  }

  /**
   * @inheritdoc
   */
  protected formatDatetime(datetime: DateTime) {
    const timestamp = super.formatDatetime(datetime)

    if (this.shouldDecorate()) {
      return chalk.dim(timestamp)
    } else {
      return timestamp
    }
  }

  /**
   * @inheritdoc
   */
  protected formatLevel(level: LogLevel) {
    const formatted = super.formatLevel(level)

    if (!this.shouldDecorate()) {
      return formatted
    } else {
      let color: string
      switch (level) {
        case 'emergency':
          color = 'red'
          break

        case 'alert':
          color = 'red'
          break

        case 'critical':
          color = 'red'
          break

        case 'error':
          color = 'red'
          break

        case 'warning':
          color = 'yellow'
          break

        case 'notice':
          color = 'blue'
          break

        case 'info':
          color = 'blue'
          break

        case 'debug':
          color = 'blue.dim'
          break

        /* c8 ignore next 2: This should never happen, but is included for type safety */
        default:
          return formatted
      }

      return chalkTemplate`{${color} ${formatted}}`
    }
  }

  /**
   * @inheritdoc
   */
  protected formatData(object: any, ignoreEmpty: boolean) {
    if (isEmpty(object) && ignoreEmpty) {
      return ''
    }

    const stringified = super.formatData(object)

    if (this.shouldDecorate()) {
      return emphasize.highlight('json', stringified).value
    } else {
      return stringified
    }
  }

  /**
   * Check whether the formatter should use ANSI codes to decorate log records
   */
  public shouldDecorate() {
    return Boolean(
      this.decorated !== undefined
        ? this.decorated
        : supportsColor
        ? supportsColor.hasBasic
        : false,
    )
  }
}
