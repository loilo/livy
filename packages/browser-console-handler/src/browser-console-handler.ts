import type { LogLevel, LogRecord, SyncHandlerInterface } from '@livy/contracts'
import { SeverityMap } from '@livy/contracts'
import {
  AbstractLevelBubbleHandler,
  AbstractLevelBubbleHandlerOptions,
} from '@livy/util/handlers/abstract-level-bubble-handler'
import { MirrorSyncHandlerMixin } from '@livy/util/handlers/mirror-sync-handler-mixin'
import { ProcessableHandlerMixin } from '@livy/util/handlers/processable-handler-mixin'
import { isEmpty } from '@livy/util/helpers'
import type { DateTime } from 'luxon'

export interface BrowserConsoleHandlerOptions
  extends AbstractLevelBubbleHandlerOptions {
  /**
   * Whether to use the browser's built-in console.debug() for the "debug" level,
   * which is only visible in the dev tools when explicitly configured so
   */
  useNativeDebug: boolean

  /**
   * Whether to include timestamps in the output
   */
  timestamps: boolean

  /**
   * The Console object to use
   */
  console: Console
}

/**
 * Writes log records to a browser console
 */
export class BrowserConsoleHandler
  extends MirrorSyncHandlerMixin(
    ProcessableHandlerMixin(AbstractLevelBubbleHandler),
  )
  implements SyncHandlerInterface
{
  private useNativeDebug: boolean
  private timestamps: boolean
  private console: Console

  public constructor({
    useNativeDebug = false,
    timestamps = false,
    console = self.console,
    ...options
  }: Partial<BrowserConsoleHandlerOptions> = {}) {
    super(options)

    this.console = console
    this.timestamps = timestamps
    this.useNativeDebug = useNativeDebug
  }

  /**
   * @inheritdoc
   */
  protected formatDatetime(datetime: DateTime) {
    return ['color: #888', datetime.toFormat('HH:mm:ss.SSS')]
  }

  /**
   * @inheritdoc
   */
  protected formatLevel(level: LogLevel) {
    const formatted = level.toUpperCase()

    let color: string
    switch (level) {
      case 'emergency':
        color = 'background-color: #f43a63; padding: 0 0.4em; color: #fff'
        break

      case 'alert':
        color = 'background-color: #f43a63; padding: 0 0.4em; color: #fff'
        break

      case 'critical':
        color = 'color: #f43a63'
        break

      case 'error':
        color = 'color: #f43a63'
        break

      case 'warning':
        color = 'color: #ffaa2b'
        break

      case 'notice':
        color = 'color: rgb(60, 161, 224)'
        break

      case 'info':
        color = 'color: rgb(60, 161, 224)'
        break

      case 'debug':
      default:
        color = 'color: rgba(60, 161, 224, 0.7)'
        break
    }

    return [color, formatted]
  }

  /**
   * @inheritdoc
   */
  public handleSync(record: LogRecord) {
    if (!this.isHandling(record.level)) {
      return
    }

    const resetter = 'color: inherit; background-color: inherit'
    const parameters: any[] = [record.message]
    const hasExtra = !isEmpty(record.extra)

    if (hasExtra || !isEmpty(record.context)) {
      parameters.push(record.context)
    }

    if (hasExtra) {
      parameters.push(record.extra)
    }

    const formattedTimestamp = this.timestamps
      ? [...this.formatDatetime(record.datetime), resetter]
      : []
    const formattedLevel = [...this.formatLevel(record.level), resetter]

    let placeholders = ['%c%s']
    if (this.timestamps) {
      placeholders.push('%c%s')
    }

    placeholders = placeholders.map(placeholder => `${placeholder}%c`)

    if (record.severity <= SeverityMap.error) {
      this.console.error(
        placeholders.join(' '),
        ...formattedTimestamp,
        ...formattedLevel,
        ...parameters,
      )
    } else if (record.severity === SeverityMap.warning) {
      this.console.warn(
        placeholders.join(' '),
        ...formattedTimestamp,
        ...formattedLevel,
        ...parameters,
      )
    } else if (record.severity <= SeverityMap.info) {
      this.console.info(
        placeholders.join(' '),
        ...formattedTimestamp,
        ...formattedLevel,
        ...parameters,
      )
    } else {
      if (this.useNativeDebug) {
        placeholders.pop()

        this.console.debug(
          placeholders.join(' '),
          ...formattedTimestamp,
          ...parameters,
        )
      } else {
        this.console.log(
          placeholders.join(' '),
          ...formattedTimestamp,
          ...formattedLevel,
          ...parameters,
        )
      }
    }

    return !this.bubble
  }
}
