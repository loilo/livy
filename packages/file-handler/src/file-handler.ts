import * as fs from 'node:fs'
import { dirname } from 'node:path'
import type {
  ClosableHandlerInterface,
  FormatterInterface,
  LogRecord,
} from '@livy/contracts'
import { EOL } from '@livy/util/environment'
import { AbstractSyncFormattingProcessingHandler } from '@livy/util/handlers/abstract-formatting-processing-handler'
import { AbstractLevelBubbleHandlerOptions } from '@livy/util/handlers/abstract-level-bubble-handler'

export interface FileHandlerOptions extends AbstractLevelBubbleHandlerOptions {
  /**
   * The formatter to use
   */
  formatter: FormatterInterface

  /**
   * Prepend one or more lines to the file when it's first created
   * This can be useful to prepend header data, like in CSV
   */
  prefix: string | string[]

  /**
   * Write the file in append mode (i.e. don't erase previous contents)
   */
  append: boolean
}

/**
 * Writes log records to a file
 */
export class FileHandler
  extends AbstractSyncFormattingProcessingHandler
  implements ClosableHandlerInterface
{
  protected fileHandle?: number
  protected path: string
  protected append: boolean
  protected prefix: string[]

  public constructor(
    path: string,
    {
      formatter,
      prefix = [],
      append = true,
      ...options
    }: Partial<FileHandlerOptions> = {},
  ) {
    super(options)

    this.append = append
    this.prefix = typeof prefix === 'string' ? [prefix] : prefix
    this.explicitFormatter = formatter

    const directory = dirname(path)
    let stat
    try {
      stat = fs.statSync(directory)
    } catch (error) {
      throw new Error(
        `Provided log path directory "${directory}" does not exist: ${
          /* c8 ignore next: Unfortunately, code coverage somehow does not recognize this line as covered */
          error instanceof Error ? error.message : error
        }`,
      )
    }
    if (!stat.isDirectory()) {
      throw new Error(
        `Provided log path parent "${directory}" is not a directory`,
      )
    }

    this.path = path
  }

  /**
   * @inheritdoc
   */
  public close() {
    if (typeof this.fileHandle === 'number') {
      fs.closeSync(this.fileHandle)
      this.fileHandle = undefined
    }
  }

  /**
   * Get the prefix string to write
   */
  private get prefixString() {
    return this.prefix.map(line => `${line}${EOL}`).join('')
  }

  /**
   * Write the formatted record to the file
   */
  protected async write(_record: LogRecord, formatted: string) {
    // No handle to the target file exists yet
    if (typeof this.fileHandle !== 'number') {
      const needsPrefix = this.prefix.length > 0 && !fs.existsSync(this.path)

      // Get a handle to the file to write
      this.fileHandle = await new Promise((resolve, reject) => {
        fs.open(this.path, this.append ? 'a' : 'w', (error, handle) => {
          if (error) {
            reject(error)
          } else {
            resolve(handle)
          }
        })
      })

      // Write the prefix to the handle
      if (needsPrefix) {
        await new Promise((resolve, reject) => {
          fs.write(this.fileHandle!, this.prefixString, (error, result) => {
            if (error) {
              reject(error)
            } else {
              resolve(result)
            }
          })
        })
      }
    }

    await new Promise((resolve, reject) => {
      fs.write(this.fileHandle!, `${formatted}${EOL}`, (error, result) => {
        if (error) {
          reject(error)
        } else {
          resolve(result)
        }
      })
    })
  }

  /**
   * Write the formatted record to the file
   */
  protected writeSync(_record: LogRecord, formatted: string) {
    if (typeof this.fileHandle !== 'number') {
      const needsPrefix = this.prefix.length > 0 && !fs.existsSync(this.path)

      this.fileHandle = fs.openSync(this.path, this.append ? 'a' : 'w')

      if (needsPrefix) {
        fs.writeSync(this.fileHandle, this.prefixString)
      }
    }

    fs.writeSync(this.fileHandle, `${formatted}${EOL}`)
  }
}
