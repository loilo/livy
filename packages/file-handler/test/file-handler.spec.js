import {
  describe,
  expect,
  it,
  beforeAll,
  beforeEach,
  afterAll,
  afterEach,
  vi
} from 'vitest'
import * as fs from 'node:fs'
import { FileHandler } from '../src/file-handler'

const { record } = livyTestGlobals

describe('@livy/file-handler', () => {
  afterEach(() => {
    fs.__reset()
  })

  it('should write logs to the configured file', () => {
    const handler = new FileHandler('/logfile.txt')

    handler.handleSync(record('debug', 'Test FileHandler'))
    handler.handleSync(record('info', 'Test FileHandler'))
    handler.handleSync(record('notice', 'Test FileHandler'))
    handler.handleSync(record('warning', 'Test FileHandler'))
    handler.handleSync(record('error', 'Test FileHandler'))
    handler.handleSync(record('critical', 'Test FileHandler'))
    handler.handleSync(record('alert', 'Test FileHandler'))
    handler.handleSync(record('emergency', 'Test FileHandler'))

    expect(fs.toJSON()['/logfile.txt']).toMatchSnapshot()
  })

  it('should write logs to the configured file (async)', async () => {
    const handler = new FileHandler('/logfile.txt')

    await handler.handle(record('debug', 'Test FileHandler'))
    await handler.handle(record('info', 'Test FileHandler'))
    await handler.handle(record('notice', 'Test FileHandler'))
    await handler.handle(record('warning', 'Test FileHandler'))
    await handler.handle(record('error', 'Test FileHandler'))
    await handler.handle(record('critical', 'Test FileHandler'))
    await handler.handle(record('alert', 'Test FileHandler'))
    await handler.handle(record('emergency', 'Test FileHandler'))

    expect(fs.toJSON()['/logfile.txt']).toMatchSnapshot()
  })

  it('should fail when trying to write to a non-existing parent folder', () => {
    expect(() => new FileHandler('/foo/logfile.txt')).toThrow(
      /^Provided log path directory "\/foo" does not exist: /
    )
  })

  it('should fail when trying to write to an existing non-folder', () => {
    expect(() => {
      fs.writeFileSync('/foo', '')
      new FileHandler('/foo/logfile.txt')
    }).toThrowError(
      new Error(`Provided log path parent "/foo" is not a directory`)
    )
  })

  describe('needs a mock on fs.closeSync()', () => {
    let originalCloseSync
    beforeAll(() => {
      originalCloseSync = fs.closeSync
    })

    beforeEach(() => {
      fs.closeSync = vi.fn()
    })

    afterAll(() => {
      fs.closeSync = originalCloseSync
    })

    it('should close the open file handle if "close" is invoked', () => {
      const handler = new FileHandler('/logfile.txt')

      // Handle a record to open a file handle
      handler.handleSync(record('debug', 'Test FileHandler'))

      handler.close()

      expect(fs.closeSync).toHaveBeenCalledTimes(1)
      expect(fs.closeSync).toHaveBeenLastCalledWith(expect.any(Number))
    })

    it('should not attempt to close a file if none is opened', () => {
      const handler = new FileHandler('/logfile.txt')

      handler.close()

      expect(fs.closeSync).not.toHaveBeenCalled()
    })
  })

  describe('needs a mock on fs.write()', () => {
    let originalWrite

    beforeAll(() => {
      originalWrite = fs.write
    })

    beforeEach(() => {
      fs.write = vi.fn((_target, _content, callback) =>
        callback(new Error(`Mock write error`))
      )
    })

    afterAll(() => {
      fs.write = originalWrite
    })

    it('should fail trying write the prefix when fs.write() fails', async () => {
      const handler = new FileHandler('/logfile.txt', {
        prefix: ['prefix line']
      })

      let writeError
      try {
        await handler.handle(record('debug', 'Test FileHandler'))
      } catch (error) {
        writeError = error
      }

      expect(writeError).toEqual(new Error(`Mock write error`))
    })

    it('should fail trying write the log record when fs.write() fails', async () => {
      const handler = new FileHandler('/logfile.txt')

      let writeError
      try {
        await handler.handle(record('debug', 'Test FileHandler'))
      } catch (error) {
        writeError = error
      }

      expect(writeError).toEqual(new Error(`Mock write error`))
    })
  })

  it('should fail when trying to write to directory instead of file', () => {
    fs.writeFileSync('/logfile.txt', '')
    const handler = new FileHandler('/logfile.txt')
    fs.unlinkSync('/logfile.txt')
    fs.mkdirSync('/logfile.txt')

    expect(() => {
      handler.handleSync(record('debug', 'Test FileHandler'))
    }).toThrow(
      new Error(`EISDIR: illegal operation on a directory, open '/logfile.txt'`)
    )
  })

  it('should fail when trying to write to directory instead of file (async)', async () => {
    fs.writeFileSync('/logfile.txt', '')
    const handler = new FileHandler('/logfile.txt')
    fs.unlinkSync('/logfile.txt')
    fs.mkdirSync('/logfile.txt')

    let writeError
    try {
      await handler.handle(record('debug', 'Test FileHandler'))
    } catch (error) {
      writeError = error
    }

    expect(writeError).toEqual(
      new Error(`EISDIR: illegal operation on a directory, open '/logfile.txt'`)
    )
  })

  it('should append to existing log file', () => {
    fs.writeFileSync('/logfile.txt', 'pre-existing content\n')

    const handler = new FileHandler('/logfile.txt')

    handler.handleSync(record('debug', 'Test FileHandler'))

    const virtualLogfileContents = fs.toJSON()['/logfile.txt']
    expect(virtualLogfileContents).toMatch(/^pre-existing content/)
    expect(virtualLogfileContents).toMatchSnapshot()
  })

  it('should rewrite file if the "append" option is set to false', () => {
    fs.writeFileSync('/logfile.txt', 'pre-existing content\n')

    const handler = new FileHandler('/logfile.txt', { append: false })

    handler.handleSync(record('debug', 'Test FileHandler'))

    const virtualLogfileContents = fs.toJSON()['/logfile.txt']
    expect(virtualLogfileContents).not.toMatch(/^pre-existing content/)
    expect(virtualLogfileContents).toMatchSnapshot()
  })

  it('should respect the "formatter" option', () => {
    const handler = new FileHandler('/logfile.txt', {
      formatter: {
        format(record) {
          return record.level
        },
        formatBatch(records) {
          return records.map(record => this.format(record)).join('\n')
        }
      }
    })

    handler.handleSync(record('debug', 'Test FileHandler'))
    handler.handleSync(record('info', 'Test FileHandler'))
    handler.handleSync(record('notice', 'Test FileHandler'))

    const virtualLogfileContents = fs.toJSON()['/logfile.txt']
    expect(virtualLogfileContents).toBe('debug\ninfo\nnotice\n')
  })

  it('should respect the "prefix" option', () => {
    const handler = new FileHandler('/logfile.txt', {
      prefix: 'first line'
    })

    handler.handleSync(record('info'))
    handler.handleSync(record('notice'))

    expect(fs.toJSON()['/logfile.txt']).toMatchSnapshot()
  })

  it('should respect the "prefix" option (async)', async () => {
    const handler = new FileHandler('/logfile.txt', {
      prefix: 'first line'
    })

    await handler.handle(record('info'))
    await handler.handle(record('notice'))

    expect(fs.toJSON()['/logfile.txt']).toMatchSnapshot()
  })

  it('should respect the "level" option', () => {
    const handler = new FileHandler('/logfile.txt', {
      level: 'notice'
    })

    expect(handler.isHandling('info')).toBe(false)
    expect(handler.isHandling('notice')).toBe(true)

    handler.handleSync(record('info'))
    handler.handleSync(record('notice'))

    expect(fs.toJSON()['/logfile.txt']).toMatchSnapshot()
  })

  it('should respect the "bubble" option', () => {
    const bubblingHandler = new FileHandler('/logfile.txt')
    const nonBubblingHandler = new FileHandler('/logfile.txt', {
      bubble: false
    })

    expect(bubblingHandler.handleSync(record('debug'))).toBe(false)
    expect(nonBubblingHandler.handleSync(record('debug'))).toBe(true)
  })
})
