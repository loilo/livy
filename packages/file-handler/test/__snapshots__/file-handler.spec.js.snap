// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@livy/file-handler > should append to existing log file 1`] = `
"pre-existing content
2015-08-15 14:30:00 DEBUG Test FileHandler {}
"
`;

exports[`@livy/file-handler > should respect the "level" option 1`] = `
"2015-08-15 14:30:00 NOTICE  {}
"
`;

exports[`@livy/file-handler > should respect the "prefix" option (async) 1`] = `
"first line
2015-08-15 14:30:00 INFO  {}
2015-08-15 14:30:00 NOTICE  {}
"
`;

exports[`@livy/file-handler > should respect the "prefix" option 1`] = `
"first line
2015-08-15 14:30:00 INFO  {}
2015-08-15 14:30:00 NOTICE  {}
"
`;

exports[`@livy/file-handler > should rewrite file if the "append" option is set to false 1`] = `
"2015-08-15 14:30:00 DEBUG Test FileHandler {}
"
`;

exports[`@livy/file-handler > should write logs to the configured file (async) 1`] = `
"2015-08-15 14:30:00 DEBUG Test FileHandler {}
2015-08-15 14:30:00 INFO Test FileHandler {}
2015-08-15 14:30:00 NOTICE Test FileHandler {}
2015-08-15 14:30:00 WARNING Test FileHandler {}
2015-08-15 14:30:00 ERROR Test FileHandler {}
2015-08-15 14:30:00 CRITICAL Test FileHandler {}
2015-08-15 14:30:00 ALERT Test FileHandler {}
2015-08-15 14:30:00 EMERGENCY Test FileHandler {}
"
`;

exports[`@livy/file-handler > should write logs to the configured file 1`] = `
"2015-08-15 14:30:00 DEBUG Test FileHandler {}
2015-08-15 14:30:00 INFO Test FileHandler {}
2015-08-15 14:30:00 NOTICE Test FileHandler {}
2015-08-15 14:30:00 WARNING Test FileHandler {}
2015-08-15 14:30:00 ERROR Test FileHandler {}
2015-08-15 14:30:00 CRITICAL Test FileHandler {}
2015-08-15 14:30:00 ALERT Test FileHandler {}
2015-08-15 14:30:00 EMERGENCY Test FileHandler {}
"
`;
